Add New Best Practices Article On API Rate Limits (#200)
cobalthq/cobalt-product-public-docs: content/en/BestPractices/protect-your-apis-with-rate-limiting.md

> ---

Author: Would you prefer `/api-rate-limiting`?

> ---

Reviewer: That works!

> ---

Author: Done.

---

{{% pageinfo %}}
  To protect your APIs against abuse, you should implement rate limiting properly. Rate limiting is often an overlooked area. In this article, we will explore more into why you should prioritize it today rather than in the future.
{{% /pageinfo %}}

## Description
  
According to the OWASP API Top 10, lack of API rate limiting (Lack of resources and rate limiting) accounts for the 4th most common type of vulnerability in APIs. Attackers are always on the lookout for APIs with weak protection. The impact is not always limited to DoS (Denial of Service). For instance, the attacker can misuse a lack of rate limit to make phone calls to premium phone numbers. While doing so, they can easily cause a lot of financial losses to the business.

To protect your APIs against abuse, you should implement rate limiting properly. Rate limiting is often an overlooked area. In this article, we will explore more into why you should prioritize it today rather than in the future. 

> 

Author: yes, I initially though those were like metadata and separate from the article body 👍

Let’s dig deeper into some abuse scenarios to understand this:

> ## Case Study: Abuse Of Phone Calling API For Toll Fraud

Author: Hmm, in this article I have discussed toll fraud (which is limited to calling). So I will prefer to skip that 😅

> ## Case Study: Abuse Of Phone Calling API For Toll Fraud

Reviewer: I'm good with that.

  
### Background
  

> APIs may make use of phone calls for verification purposes in an application. So, how can an attacker abuse it? 

Author: in this particular case, afaik only calling is abused and not texting

> Fraudsters often find creative ways to make money off security loopholes. They usually look for websites with lax security measures. The attackers start by locating a phone verification API endpoint in your application without rate limits. Then they quickly spring into action.

Author: Hi Mike,

You can analyze the API requests using the browser's network tab, or using the frontend code both of which are readily accessible to anyone.

### Real-World Example
  

> Recently, a colleague told me how a company booked a loss of over $300k in an incident of toll fraud. This figure would only increase when your business is more valuable.

Author: Thanks, @mjang-cobalt. 

> ## Case Study: Abuse Of Phone Calling API For Toll Fraud

Initially, I called it toll fraud. I think that's a more accurate term. The second sentence with reference looks good!

You can reduce their odds of winning by rate-limiting requests hitting your API endpoints. Resultantly, the effort and cost needed for carrying out the attack increase. The attacker would need to cycle through a large pool of IP addresses to get around the rate limits. Hence, it makes your API a rather uninteresting target for them. 

While you might also need other advanced measures for fraud detection, strict rate limiting can go a long way in dissuading such attackers.

### Vulnerable Code
  
Consider the following snippet of code. A POST request with a phone number in the body to this API route triggers a verification phone call to that number.

What is wrong with it?

```

>   app.post('/verify-phone number',(req, res) => {
> 	  const { phone } = req.body;	
>     	  verify.call(phone)
> 	  //...
>   })

Reviewer: Based on a syntax check at jshint.com:

```suggestion
  app.post('/verify-phone number',(req, res) => {
	  const { phone } = req.body;	
    	  verify.call(phone);
	  //...
  });
```

```

Fraudsters can easily abuse this API endpoint for toll fraud due to the lack of rate limiting.

### Fix

Let’s fix that. Several npm packages can do this. But for this example, we will use `express-rate-limit`:

- Import it first. 
- Then define a custom rate limit for the API endpoint (`callLimiter`). 
- Set a window to 1 hr (60 * 60 * 1000 ms).
- Set max limits for that window as appropriate (10 attempts).
- Optionally, provide a friendly error message to the blocked user (`'Too many verification attempts, please try again after an hour'`)
- Finally, apply the `callLimiter` middleware we defined to the endpoint.

```
import rateLimit from 'express-rate-limit'

//...

> const callLimiter = rateLimit({
> 	windowMs: 60 * 60 * 1000, // 1 hour
> 	max: 10, // Limit each IP to 10 phone calls per `window` (here, per hour)
> 	message:
> 		'Too many verification attempts, please try again after an hour',
> 	standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
> })
>  
> //...
> app.post('/verify-phone-number', callLimiter, (req, res) => {
> 	const { phone } = req.body;	
>   	verify.call(phone)
> 	//...
> })

Reviewer: Based on a syntax check, using jshint.com:

```suggestion
const callLimiter = rateLimit({
	windowMs: 60 * 60 * 1000, // 1 hour
	max: 10, // Limit each IP to 10 phone calls per `window` (here, per hour)
	message:
		'Too many verification attempts, please try again after an hour',
	standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
});
 
//...
app.post('/verify-phone-number', callLimiter, (req, res) => {
	const { phone } = req.body;	
  	verify.call(phone);
	//...
});
```

```

## Other Scenarios
  
### Denial Of Service
  
Attackers may overload the API with a large number of requests. Thus, the API becomes unavailable to your users if it exceeds the API’s capacity to process requests. This affects the availability of your APIs.
  
### Bruteforce And Credential Stuffing Attacks
  

> Attackers often brute force user credentials with lists of common passwords. Similarly, attackers also abuse the lack of rate limiting to perform credential stuffing attacks. In the case of credential stuffing attacks, attackers attempt to log in to user accounts using credentials obtained from other breaches or stolen credentials.

Author: Hi @mjang-cobalt,

Here I was suggesting two different attack scenarios

- Brute force with a list of common passwords
- Credential stuffing with breached and stolen credentials

In the **Fix** section, I suggested a fix for all these scenarios i.e. API rate limiting.

> Attackers often brute force user credentials with lists of common passwords. Similarly, attackers also abuse the lack of rate limiting to perform credential stuffing attacks. In the case of credential stuffing attacks, attackers attempt to log in to user accounts using credentials obtained from other breaches or stolen credentials.

Reviewer: Ahhhh, they are different. A good way to make the point is with bullets, so maybe:

```suggestion
Attackers can use multiple methods to attack your systems:

- They can brute force user credentials with lists of common passwords. 
- They can abuse the lack of rate limiting to perform credential stuffing attacks. 
  - With credential stuffing attacks, attackers attempt to log in to user accounts using credentials obtained from other breaches or stolen credentials.
```

 
### Data Scraping
  
Companies have policies that restrict data scraping from their application. That doesn’t guarantee protection against bad actors. Bad actors try to circumvent your measures by pulling information from your APIs at a very high rate.
  
### Fix
  

> Use API rate limiting to protect against these kinds of attacks. Refer to the implementation in the previous scenario.

Author: @mjang-cobalt this is supposed to be a general fix for all the scenarios discussed above. Is it confusing? Do you suggest a better way to communicate this?

> ### Fix
>   
> Use API rate limiting to protect against these kinds of attacks. Refer to the implementation in the previous scenario.

Reviewer: Oh duh of course. Now I see what you mean by "Refer to the implementation in the previous scenario."

But we now have multiple scenarios. And "previous scenarios" is an awkward phrase. If we make "Fix" a level 2 header, we'll show that it applies to all scenarios in the article. So I'm revising my suggestion to:

```suggestion
## The Fix
  
Use API rate limiting to protect against these kinds of attacks.
```

> ## Some Considerations
>   
> Before you implement rate limiting for your APIs, also consider these factors:

Reviewer: Removing unneeded words:

```suggestion
## Considerations
  
Before you implement rate limiting for your APIs, consider these factors:
```

  
### Will It Impact User Experience?
  

> Well, it depends. You might want to give the user more freedom when using certain API functionalities. On the other hand, you might want to be stricter for abuse-prone APIs. This requires planning. Plan and tailor your rate limits to your use case before applying them. Choose the limits based on how many requests you expect from the client applications under normal conditions.

Reviewer: @sourcecodereviewer can you include an example of when it makes sense to "give the user more freedom"

I think we can then split this section into 2 or even 3 paragraphs

Author: sure @mjang-cobalt 👍

> ### Is Rate Limit Sufficient Alone?
>   
> While application layer rate limiting is a very good practice, it might not be sufficient alone if your attackers are more sophisticated. In some cases, attackers can outpace you. They quickly rotate IPs from large numbers of IP addresses belonging to real users that are part of a botnet. In such cases, blocking bad reputation IP addresses from threat intelligence feeds could be a more efficient way.

Reviewer: I suggest rephrasing:

```suggestion
### Going Further
  
While application layer rate limiting is a good practice, you may need to do more. Sophisticated attackers may
set up attacks from a large numbers of IP addresses. Some botnets include IP addresses from real users.
In response, you could review threat intelligence feeds to block bad reputation IP addresses.
```

> Furthermore, depending on your use case you might need more advanced anti-fraud measures and geo-restrictions apart from API rate limiting.

Reviewer: ```suggestion
Furthermore, depending on your use case you might need more advanced anti-fraud measures and geo-restrictions.

```
